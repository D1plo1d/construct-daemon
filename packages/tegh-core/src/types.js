/* config */
export Config, { MockConfig } from './config/types/Config'
export * as AxisTypeEnum from './config/types/AxisTypeEnum'

/* jobQueue */
export * as JobHistoryTypeEnum from './jobQueue/types/JobHistoryTypeEnum'

export JobHistoryEvent from './jobQueue/types/JobHistoryEvent'
export JobQueueGraphQL from './jobQueue/types/JobQueue.graphql.js'
export Job, { MockJob } from './jobQueue/types/Job'
export JobGraphQL from './jobQueue/types/Job.graphql.js'
export JobFile, { MockJobFile } from './jobQueue/types/JobFile'

/* printer */
export PrinterGraphQL from './printer/types/Printer.graphql.js'

/* spool */
export * as PriorityEnum from './spool/types/PriorityEnum'
export * as TaskStatusEnum from './spool/types/TaskStatusEnum'
export * as ComponentTypeEnum from './config/types/ComponentTypeEnum'

export Task, { MockTask } from './spool/types/Task'
export TaskGraphQL from './spool/types/Task.graphql.js'
