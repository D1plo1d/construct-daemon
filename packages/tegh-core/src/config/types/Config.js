import { Record, List } from 'immutable'
import uuid from 'uuid/v4'

import HostConfig from './HostConfig'
import PrinterConfig from './PrinterConfig'
import MaterialConfig from './MaterialConfig'
import developmentConfig from '../../../../../development.config'

export const ConfigRecordFactory = Record({
  id: null,
  modelVersion: 0,
  host: null,
  printer: null,
  materials: List(),
})

const Config = ({
  id = uuid(),
  modelVersion = 0,
  host = {},
  printer = {},
  materials = [],
  ...props
} = {}) => (
  ConfigRecordFactory({
    id,
    modelVersion,
    host: HostConfig(host),
    printer: PrinterConfig(printer),
    materials: List(materials).map(material => MaterialConfig(material)),
    ...props,
  })
)

export const MockConfig = ({
  host = {},
  printer = {},
  ...props
} = {}) => (
  Config({
    ...developmentConfig,
    host: {
      ...developmentConfig.host,
      ...host,
    },
    printer: {
      ...developmentConfig.printer,
      ...printer,
    },
    ...props,
  })
)

export default Config
