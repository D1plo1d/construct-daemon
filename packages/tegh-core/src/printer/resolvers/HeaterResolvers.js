import getComponentsState from '../selectors/getComponentsState'
import getHeaterMaterialTargets from '../selectors/getHeaterMaterialTargets'

const HeaterResolvers = {
  Heater: {
    history: (source, args, { store }) => {
      const state = store.getState()
      return getComponentsState(state).temperatureHistory
        .filter(entry => entry.componentID === source.id)
        .toArray()
    },
    materialTarget: (source, args, { store }) => {
      const { config } = store.getState()

      return getHeaterMaterialTargets(config).get(source.id)
    },
  },
}

export default HeaterResolvers
