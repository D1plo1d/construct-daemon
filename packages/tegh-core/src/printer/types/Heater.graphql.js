import tql from 'typiql'
import snl from 'strip-newlines'
import {
  GraphQLObjectType,
} from 'graphql'

import getHeaterConfigs from '../../config/selectors/getHeaterConfigs'

const HeaterType = new GraphQLObjectType({
  name: 'Heater',
  description: 'A heated bed or extruder',
  fields: () => ({
    id: {
      type: tql`ID!`,
      resolve: source => `${source.id}DynamicData`,
    },
    name: {
      type: tql`String!`,
      resolve: (source, args, { store }) => {
        const state = store.getState()
        return getHeaterConfigs(state.config).get(source.id).name
      },
    },
    type: {
      type: tql`String!`,
      resolve: (source, args, { store }) => {
        const state = store.getState()
        return getHeaterConfigs(state.config).get(source.id).type
      },
    },
    targetTemperature: {
      type: tql`Float`,
      description: snl`
        The target temperature in °C for this heater. The heater will
        attempt to make the current_temp equal to this temperature.
      `,
    },
    currentTemperature: {
      type: tql`Float!`,
      description: snl`
        The current temperature in °C recorded by the heater’s thermocouple
        or thermister.
      `,
    },
    blocking: {
      type: tql`Boolean!`,
      description: snl`
        True if the printer is waiting on this heater to reach it’s
        targetTemp and preventing any more gcodes from executing until it does.
      `,
    },
  }),
})

export default HeaterType
