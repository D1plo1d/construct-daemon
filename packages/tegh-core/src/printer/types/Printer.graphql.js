import { GraphQLObjectType } from 'graphql'
import tql from 'typiql'
import snl from 'strip-newlines'

import isIdle from '../../spool/selectors/isIdle'
import getDriverState from '../selectors/getDriverState'

import PrinterStatusEnum from './PrinterStatusEnum.graphql'
import HeaterType from './Heater.graphql'
import FanType from './Fan.graphql'
import LogEntryType from '../../log/types/LogEntry.graphql'
import MacroDefinitionType from '../../config/types/MacroDefinition.graphql'
import PrinterErrorType from './PrinterError.graphql'

const PrinterGraphQL = new GraphQLObjectType({
  name: 'Printer',
  fields: () => ({
    id: {
      type: tql`ID!`,
      resolve(source) {
        return source.config.id
      },
    },
    name: {
      type: tql`String!`,
      resolve(source) {
        return source.config.name
      },
    },
    heaters: {
      type: tql`[${HeaterType}!]!`,
      resolve(source) {
        return getDriverState(source).peripherals.heaters.values()
      },
    },
    targetTemperaturesCountdown: {
      type: tql`Float`,
      description: snl`
        The estimated number of seconds until the heater(s) reach their
        targetTemperature.
      `,
      resolve(source) {
        return getDriverState(source).peripherals.targetTemperaturesCountdown
      },
    },
    activeExtruderID: {
      type: tql`String`,
      description: snl`
        The active extruder ID
      `,
      resolve(source) {
        return getDriverState(source).peripherals.activeExtruderID
      },
    },
    fans: {
      type: tql`[${FanType}!]!`,
      resolve(source) {
        return getDriverState(source).peripherals.fans.values()
      },
    },
    status: {
      type: tql`${PrinterStatusEnum}!`,
      resolve: (source) => {
        if (!isIdle(source.spool)) return 'PRINTING'
        const { status } = source.status
        return status.substring(status.lastIndexOf('/') + 1)
      },
    },
    error: {
      type: tql`${PrinterErrorType}`,
      resolve: source => source.status.error,
    },
    macroDefinitions: {
      type: tql`[${MacroDefinitionType}!]!`,
      resolve: source => source.macros.keys(),
    },
    logEntries: {
      type: tql`[${LogEntryType}!]`,
      args: {
        level: {
          type: tql`String`,
        },
        source: {
          type: tql`String`,
        },
      },
      resolve(source, args) {
        let entries = source.log.get('entries')
        if (args.level != null) {
          entries = entries.filter(log => log.level === args.level)
        }
        if (args.source != null) {
          entries = entries.filter(log => log.source === args.source)
        }
        return entries.toArray()
      },
    },
  }),
})

export default PrinterGraphQL
