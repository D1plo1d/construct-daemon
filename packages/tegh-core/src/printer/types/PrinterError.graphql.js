import {
  GraphQLObjectType,
} from 'graphql'
import tql from 'typiql'
import snl from 'strip-newlines'

const PrinterErrorType = new GraphQLObjectType({
  name: 'PrinterError',
  fields: () => ({
    code: {
      type: tql`String!`,
      description: snl`
        A machine-readable code indicating the type of error
      `,
    },
    message: {
      type: tql`String!`,
      description: snl`
        A human-readable description of the error
      `,
    },
  }),
})

export default PrinterErrorType
