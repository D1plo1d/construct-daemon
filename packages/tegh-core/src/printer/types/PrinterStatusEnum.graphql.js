import snl from 'strip-newlines'
import {
  GraphQLEnumType,
} from 'graphql'

const PrinterStatusEnum = new GraphQLEnumType({
  name: 'PrinterStatusEnum',
  values: {
    CONNECTING: {
      description: snl`
        The printer is being initialized.
        Attempting to spool anything will result in an error.
      `,
    },
    READY: {
      description: snl`
        The printer is connected and able to spool tasks/jobs.
      `,
    },
    PRINTING: {
      description: snl`
        The printer is printing a job.
        Attempting to spool anything will result in an error.
      `,
    },
    DISCONNECTED: {
      description: snl`
        The printer is disconnected or turned off.
        Attempting to spool anything will result in an error.
      `,
    },
    ERRORED: {
      description: snl`
        The printer is being initialized.
        Attempting to spool anything except for an
        emergency macro (ie. reset) will result in an error. Spool the \`reset\`
        macro to reset the error and change the status to \`CONNECTING\`.
      `,
    },
    ESTOPPED: {
      description: snl`
        The printer is estopped. Attempting to spool anything except for an
        emergency macro (ie. reset) will result in an error. Spool the \`reset\`
        macro to reset the estop and change the status to \`CONNECTING\`.
      `,
    },
  },
})

export default PrinterStatusEnum
