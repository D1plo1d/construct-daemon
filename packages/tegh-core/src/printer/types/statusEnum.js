export const ERRORED = 'tegh/printer/status/ERRORED'
export const ESTOPPED = 'tegh/printer/status/ESTOPPED'
export const DISCONNECTED = 'tegh/printer/status/DISCONNECTED'
export const CONNECTING = 'tegh/printer/status/CONNECTING'
export const READY = 'tegh/printer/status/READY'
