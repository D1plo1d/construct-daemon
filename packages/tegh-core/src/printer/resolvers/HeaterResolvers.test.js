import { Record } from 'immutable'
import typeDefs from 'tegh-schema'

import deterministicTestSetup from '../../util/testing/deterministicTestSetup'
import snapshotTestResolvers from '../../util/testing/snapshotTestResolvers'

import { MockConfig } from '../../config/types/Config'

import HeaterResolvers from './HeaterResolvers'

describe('JobResolvers', () => {
  it('presents a consistent API', () => {
    deterministicTestSetup()

    const state = Record({
      config: MockConfig(),
    })()

    snapshotTestResolvers({
      typeDefs,
      typeName: 'Heater',
      resolvers: HeaterResolvers,
      rootValue: Record({
        id: 'heater_1',
        address: 'e1',
        currentTemperature: 32,
        targetTemperature: 40,
        blocking: true,
      })(),
      contextValue: {
        store: {
          getState: () => state,
        },
      },
    })
  })
})
