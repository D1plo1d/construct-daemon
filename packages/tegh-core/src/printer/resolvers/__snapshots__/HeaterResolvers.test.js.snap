// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaterResolvers presents a consistent API 1`] = `
Object {
  "_testHeater": Object {
    "blocking": true,
    "currentTemperature": 32,
    "id": "heater_1",
    "name": "Extruder 1",
    "targetTemperature": 40,
    "type": "TOOLHEAD",
  },
}
`;
